File: MachineLearningAZ/Part01_Preprocessing/data_preprocessing_template_JEC.py
# Template para preprocesar datos 

# Importamos librerias 
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# esto es para ver en que directorio estamos
import os;
print os.getcwd();

# Importamos el dataset
dataset = pd.read_csv('Data.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 3].values

# Tratamos los 'missing values' con el imputer
from sklearn.preprocessing import Imputer

imputer = Imputer(missing_values='NaN', strategy='mean', axis=0)
# If “mean”, then replace missing values using the mean along the axis.
# If “median”, then replace missing values using the median along the axis.
# If “most_frequent”, then replace missing using the most frequent value along the axis.
imputer = imputer.fit(X[:,1:3])
X[:,1:3]=imputer.transform(X[:,1:3])


# Codificamos los datos categoricos
#-----------------------------------
# Tratamos las variables independientes 
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X = LabelEncoder()
X[:, 0] = labelencoder_X.fit_transform(X[:, 0])
onehotencoder = OneHotEncoder(categorical_features = [0])
X = onehotencoder.fit_transform(X).toarray()

# Tratamos la variable objetivo
labelencoder_y = LabelEncoder()
y = labelencoder_y.fit_transform(y)


# Dividimos los datos en Training set y Test set
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

# Escalamos 
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

File: MachineLearningAZ/Part01_Preprocessing/data_preprocessing_templateJEC.R
# Data Preprocessing en R

# Importamos el dataset
dataset = read.csv('Data.csv')
head(dataset)
# dataset = dataset[, 2:3]

# Dividimos el dataset en Training y Test con caTools
# install.packages('caTools')
library(caTools)
set.seed(123)

# Tratamos los datos faltantes con la media
# En este caso Age y Salary 
dataset$Age = ifelse(is.na(dataset$Age),
                     ave(dataset$Age, FUN = function(x) mean(x, na.rm = TRUE)),
                     dataset$Age)
dataset$Salary = ifelse(is.na(dataset$Salary),
                        ave(dataset$Salary, FUN = function(x) mean(x, na.rm = TRUE)),
                        dataset$Salary)

# Codificamos los datos categoricos
dataset$Country = factor(dataset$Country,
                         levels = c('France', 'Spain', 'Germany'),
                         labels = c(1, 2, 3))
dataset$Purchased = factor(dataset$Purchased,
                           levels = c('No', 'Yes'),
                           labels = c(0, 1))

# Con SplitRatio establecemos el porcentaje de separacion
split = sample.split(dataset$Purchased, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Escalado
training_set[, 2:3] = scale(training_set[, 2:3])
test_set[, 2:3] = scale(test_set[, 2:3])
